#' Estimation Parameters Using muscat
#'
#' @param SCST_Object A SCST object generated by [SCST::Create_SCST_Object()]
#' @param min_count,min_cells Only genes with a count > min_count in >= min_cells will be retained.
#' @param min_genes Only cells with a count > 0 in >= min_genes will be retained.
#' @param verbose Whether to return messages or not
#' @param ... Other parameters represented in muscat, see [muscat::prepSim()]
#'
#' @references Crowell H L, Soneson C, Germain P L, et al. Muscat detects subpopulation-specific state transitions from multi-sample multi-condition single-cell transcriptomics data. Nature communications, 2020, 11(1): 1-12. <https://doi.org/10.1038/s41467-020-19894-4>
#'
#' @export
#'
muscat_estimation <- function(SCST_Object,
                              min_count = 0,
                              min_cells = 0,
                              min_genes = 0,
                              verbose = FALSE,
                              ...
){
  ##############################################################################
  ####                            Environment                                ###
  ##############################################################################
  if(!requireNamespace("muscat", quietly = TRUE)){
    message("muscat is not installed on your device...")
    message("Installing muscat...")
    devtools::install_github("HelenaLC/muscat")
  }
  ##############################################################################
  ####                               Check                                   ###
  ##############################################################################
  ref_data <- SCST_Object@reference %>% as.matrix()
  seed <- SCST_Object@seed
  ## group
  if(S4Vectors::isEmpty(SCST_Object@group_label)){
    group_label <- rep("A", ncol(ref_data))
  }else{
    group_label <- SCST_Object@group_label
  }

  col_data <- data.frame("sample_id" = rep("A", ncol(ref_data)),
                         "group_id" = rep("A", ncol(ref_data)),
                         "cluster_id" = group_label)
  ref_data <- SingleCellExperiment::SingleCellExperiment(list(counts = ref_data),
                                                         colData = col_data)
  ref_data <- muscat::prepSCE(x = ref_data)
  existed_params <- list("x" = ref_data,
                         "min_count" = min_count,
                         "min_cells" = min_cells,
                         "min_genes" = min_genes,
                         "verbose" = verbose,
                         "group_keep" = levels(as.factor(col_data$"group_id")))
  existed_params <- append(existed_params, list(...))
  argument <- formals(muscat::prepSim)
  used_argument <- change_parameters(existed_params, argument)
  used_argument <- used_argument[-grep("min_size", names(used_argument))]
  used_argument <- append(used_argument, list("min_size" = NULL))
  ##############################################################################
  ####                            Estimation                                 ###
  ##############################################################################
  if(verbose){
    message("Estimating parameters using muscat")
  }
  # Seed
  set.seed(seed)
  # Estimation
  estimate_detection <- peakRAM::peakRAM(
    estimate_result <- do.call(muscat::prepSim, used_argument)
  )
  ##############################################################################
  ####                           Ouput                                       ###
  ##############################################################################
  slot(SCST_Object, "estimation_time_memory") <- list("muscat" = list("estimation_time" = estimate_detection$Elapsed_Time_sec,
                                                                      "estimation_memory" = estimate_detection$Peak_RAM_Used_MiB))
  slot(SCST_Object, "estimation_result") <- list("muscat" = estimate_result)
  slot(SCST_Object, "estimation_parameters") <- list("muscat" = used_argument[-1])
  return(SCST_Object)
}



#' Simulate ScRNA-seq Data Using muscat
#'
#' @param estimated_result The SCST object after estimating parameters using [splatter::muscat_estimation()]
#' @param cell_num The expected number of simulated cells. If NULL, the original cell number in the reference data will be adopted.
#' @param gene_num The expected number of simulated genes. If NULL, the original gene number in the reference data will be adopted.
#' @param nGroups The expected number of groups in the simulated data. Default is 1.
#' @param prop_group The proportion of cells in each group. Default is 1.
#' @param fc_group The fold change of DEGs between cell groups. Default is 2.
#' @param de_prop The proportion of DEGs over all genes between different simulated cell groups. Default is 0.2.
#' @param verbose Whether to return messages or not
#' @param seed Random seed
#' @param return_format The format of returned simulation data. Choices: list, SingleCellExperiment and Seurat.
#' @param ... Other parameters represented in muscat, see [muscat::simData()]
#'
#' @importFrom S4Vectors metadata
#' @importFrom dplyr transmute case_when
#'
#' @references Crowell H L, Soneson C, Germain P L, et al. Muscat detects subpopulation-specific state transitions from multi-sample multi-condition single-cell transcriptomics data. Nature communications, 2020, 11(1): 1-12. <https://doi.org/10.1038/s41467-020-19894-4>
#'
#' @export
#'
muscat_simulation <- function(estimated_result,
                              cell_num = NULL,
                              gene_num = NULL,
                              nGroups = 1,
                              prop_group = 1,
                              fc_group = 2,
                              de_prop = 0.2,
                              return_format,
                              verbose = FALSE,
                              seed,
                              ...
){
  ##############################################################################
  ####                            Environment                                ###
  ##############################################################################
  if(!requireNamespace("muscat", quietly = TRUE)){
    message("muscat is not installed on your device...")
    message("Installing muscat...")
    devtools::install_github("HelenaLC/muscat")
  }
  existed_params <- list("x" = estimated_result@estimation_result$muscat)
  ## nCells
  if(!is.null(cell_num)){
    existed_params[["nc"]] <- cell_num
  }else{
    existed_params[["nc"]] <- ncol(estimated_result@estimation_result$muscat)
  }
  ## nGenes
  if(!is.null(gene_num)){
    existed_params[["ng"]] <- gene_num
  }else{
    existed_params[["ng"]] <- nrow(estimated_result@estimation_result$muscat)
  }
  ## fc_group
  existed_params[["lfc"]] <- log2(fc_group)
  ## de_prop
  existed_params[["p_dd"]] <- c(1-de_prop, 0, de_prop, 0, 0, 0)
  ## nGroups
  if(nGroups == 1){
    existed_params[["dd"]] <- FALSE
    existed_params[["nk"]] <- 1
  }else{
    existed_params[["ns"]] <- 1
    existed_params[["nk"]] <- nGroups
    existed_params[["probs"]] <- list(prop_group, NULL, NULL)
    existed_params[["dd"]] <- FALSE
    existed_params[["paired"]] <- FALSE
  }
  ## phylo_pars
  existed_params[["phylo_pars"]] <- c(0, 3)
  existed_params[["force"]] <- TRUE
  existed_params <- append(existed_params, list(...))
  argument <- formals(muscat::simData)
  used_argument <- change_parameters(existed_params, argument)
  ##############################################################################
  ####                            Simulation                                 ###
  ##############################################################################
  if(verbose){
    message("Simulating datasets using muscat")
  }
  # Seed
  set.seed(seed)
  # Simulation
  simulate_detection <- peakRAM::peakRAM(
    simulate_result <- do.call(muscat::simData, used_argument)
  )
  ##############################################################################
  ####                        Format Conversion                              ###
  ##############################################################################
  ## counts
  counts <- SingleCellExperiment::counts(simulate_result)
  colnames(counts) <- paste0("Cell", 1:ncol(counts))
  rownames(counts) <- paste0("Gene", 1:nrow(counts))
  ## gene information
  gene_info <- S4Vectors::metadata(simulate_result)$gene_info %>%
    dplyr::select(c(1,2,4)) %>%
    dplyr::group_by(gene) %>%
    dplyr::summarise(logFC = sum(logFC)) %>%
    dplyr::mutate(gene = gsub("gene", "Gene", gene)) %>%
    as.data.frame()
  rownames(gene_info) <- gene_info$gene
  gene_info <- gene_info[rownames(counts), ]
  row_data <- gene_info %>%
    dplyr::transmute("gene_name" = rownames(counts),
                     "de_gene" = dplyr::case_when(
                       is.na(gene_info$"logFC") ~ "no",
                       TRUE ~ "yes"
                     ))
  rownames(row_data) <- row_data$gene_name
  ## cell information
  if(is.null(simulate_result$cluster_id)){
    group <- rep("Group1", ncol(counts))
  }else{
    group <- paste0("Group", as.numeric(simulate_result$cluster_id))
  }
  col_data <- data.frame("cell_gene" = colnames(counts),
                         "group" = group,
                         row.names = colnames(counts))
  rownames(col_data) <- col_data$cell_gene
  # Establish SingleCellExperiment
  simulate_result <- SingleCellExperiment::SingleCellExperiment(list(counts = counts),
                                                                colData = col_data,
                                                                rowData = row_data)
  simulate_result <- data_conversion(SCE_object = simulate_result, return_format = return_format)
  ##############################################################################
  ####                           Ouput                                       ###
  ##############################################################################
  slot(estimated_result, "simulation_time_memory") <- list("muscat" = list("estimation_time" = simulate_detection$Elapsed_Time_sec,
                                                                           "estimation_memory" = simulate_detection$Peak_RAM_Used_MiB))
  slot(estimated_result, "simulation_result") <- list("muscat" = simulate_result)
  slot(estimated_result, "simulation_parameters") <- list("muscat" = used_argument[-1])
  return(estimated_result)
}
