#' Estimation Parameters Using scDesign2
#'
#' @param SCST_Object A SCST object generated by [SCST::Create_SCST_Object()]
#' @param cell_type A character string specifying the cluster labels of cells.
#' @param sim_method Specification of the type of model. Default value is "copula", which selects the copula model. "ind" will select the (w/o copula) model.
#' @param marginal A string or a vector of strings of the marginal distribution. Must be one of "poisson", "nb", "zip", "zinb" or "gaussian". Default is "zinb".
#' @param verbose Whether to return messages or not
#' @param ... Other parameters represented in scDesign2, see [scDesign2::fit_model_scDesign2()]
#' @references Sun T, Song D, Li W V, et al. scDesign2: a transparent simulator that generates high-fidelity single-cell gene expression count data with gene correlations captured. Genome biology, 2021, 22(1): 1-37. <https://doi.org/10.1186/s13059-021-02367-2>
#'
#' @export
#'
#'
scDesign2_estimation <- function(SCST_Object,
                                 cell_type = NULL,
                                 sim_method = "copula",
                                 marginal = "auto_choose",
                                 verbose = FALSE,
                                 ...
){
  ##############################################################################
  ####                            Environment                                ###
  ##############################################################################
  if(!requireNamespace("scDesign2", quietly = TRUE)){
    message("scDesign2 is not installed on your device...")
    message("Installing scDesign2...")
    devtools::install_github("JSB-UCLA/scDesign2")
  }
  ##############################################################################
  ####                               Check                                   ###
  ##############################################################################
  ref_data <- as.matrix(SCST_Object@reference)
  seed <- SCST_Object@seed
  if(!is.null(cell_type)){
    colnames(ref_data) <- cell_type
    cell_type_sel <- unique(cell_type)
  }else{
    if(!S4Vectors::isEmpty(SCST_Object@group_label)){
      fake_cell_name <- paste0(as.character(SCST_Object@group_label), "cell")
      colnames(ref_data) <- fake_cell_name
      cell_type_sel <- unique(fake_cell_name)
    }else{
      colnames(ref_data) <- rep("Acell", ncol(ref_data))
      cell_type_sel <- "Acell"
    }
  }
  existed_params <- list("data_mat" = ref_data,
                         "cell_type_sel" = cell_type_sel,
                         "sim_method" = sim_method,
                         "marginal" = marginal)
  existed_params <- append(existed_params, list(...))
  argument <- formals(scDesign2::fit_model_scDesign2)
  used_argument <- change_parameters(existed_params, argument)
  ##############################################################################
  ####                            Estimation                                 ###
  ##############################################################################
  if(verbose){
    message("Estimating parameters using scDesign2")
  }
  # Seed
  set.seed(seed)
  # Estimation
  estimate_detection <- peakRAM::peakRAM(
    estimate_result <- do.call(scDesign2::fit_model_scDesign2, used_argument)
  )
  ##############################################################################
  ####                           Ouput                                       ###
  ##############################################################################
  slot(SCST_Object, "estimation_time_memory") <- list("scDesign2" = list("estimation_time" = estimate_detection$Elapsed_Time_sec,
                                                                         "estimation_memory" = estimate_detection$Peak_RAM_Used_MiB))
  slot(SCST_Object, "estimation_result") <- list("scDesign2" = estimate_result)
  slot(SCST_Object, "estimation_parameters") <- list("scDesign2" = used_argument[-1])
  return(SCST_Object)
}


#' Simulate ScRNA-seq Data Using scDesign2
#'
#' @param estimated_result The SCST object after estimating parameters using [SCST::scDesign2_estimation()]
#' @param cell_num The expected number of simulated cells. If NULL, the original cell number in the reference data will be adopted.
#' @param prop_group The proportion of cells in each group. Default is 1.
#' @param verbose Whether to return messages or not
#' @param seed Random seed
#' @param return_format The format of returned simulation data. Choices: list, SingleCellExperiment and Seurat.
#' @param ... Other parameters represented in scDesign2, see [scDesign2::simulate_count_scDesign2()]
#'
#' @references Sun T, Song D, Li W V, et al. scDesign2: a transparent simulator that generates high-fidelity single-cell gene expression count data with gene correlations captured. Genome biology, 2021, 22(1): 1-37. <https://doi.org/10.1186/s13059-021-02367-2>
#'
#' @export
#'
scDesign2_simulation <- function(estimated_result,
                                 cell_num = NULL,
                                 prop_group = 1,
                                 return_format,
                                 verbose = FALSE,
                                 seed,
                                 ...
){

  ##############################################################################
  ####                            Environment                                ###
  ##############################################################################
  if(!requireNamespace("scDesign2", quietly = TRUE)){
    message("scDesign2 is not installed on your device...")
    message("Installing scDesign2...")
    devtools::install_github("JSB-UCLA/scDesign2")
  }
  parameters <- estimated_result@estimation_result$scDesign2
  ## nCells
  if(!is.null(cell_num)){
    n_cell_new <- cell_num
  }else{
    n_cell_new <- estimated_result@cell_num
  }
  ## prob.group
  if(length(prop_group) != length(parameters)){
    cell_type_prop <- rep(1/length(parameters), length(parameters))
  }else{
    cell_type_prop <- prop_group
  }

  existed_params <- list("model_params" = parameters,
                         "n_cell_new" = n_cell_new,
                         "cell_type_prop" = cell_type_prop)
  existed_params <- append(existed_params, list(...))
  argument <- formals(scDesign2::simulate_count_scDesign2)
  used_argument <- change_parameters(existed_params, argument)
  ##############################################################################
  ####                            Simulation                                 ###
  ##############################################################################
  if(verbose){
    message("Simulating datasets using scDesign2")
  }
  # Seed
  set.seed(seed)
  # Estimation
  simulate_detection <- peakRAM::peakRAM(
    simulate_result <- BiocGenerics::do.call(scDesign2::simulate_count_scDesign2, used_argument)
  )
  ##############################################################################
  ####                        Format Conversion                              ###
  ##############################################################################
  counts <- simulate_result
  # Rename
  rownames(counts) <- paste0("Gene", 1:nrow(counts))
  colnames(counts) <- paste0("Cell", 1:ncol(counts))
  # col_data
  col_data <- data.frame("cell_name" = colnames(counts),
                         "group" = paste0("Group", rep(1:length(table(colnames(simulate_result))),
                                                       table(colnames(simulate_result)))),
                         row.names = colnames(counts))
  # Row data
  row_data <- data.frame("gene_name" = rownames(counts),
                         row.names = rownames(counts))
  # Establish SingleCellExperiment
  simulate_result <- SingleCellExperiment::SingleCellExperiment(list(counts = counts),
                                                                colData = col_data,
                                                                rowData = row_data)
  simulate_result <- data_conversion(SCE_object = simulate_result, return_format = return_format)
  ##############################################################################
  ####                           Ouput                                       ###
  ##############################################################################
  slot(estimated_result, "simulation_time_memory") <- list("scDesign2" = list("estimation_time" = simulate_detection$Elapsed_Time_sec,
                                                                              "estimation_memory" = simulate_detection$Peak_RAM_Used_MiB))
  slot(estimated_result, "simulation_result") <- list("scDesign2" = simulate_result)
  slot(estimated_result, "simulation_parameters") <- list("scDesign2" = used_argument[-1])
  return(estimated_result)
}
