#' Simulate ScRNA-seq Data Using scDesign
#'
#' @param SCST_Object The SCST object generated by [SCST::Create_SCST_Object()]
#' @param cell_num The expected number of simulated cells. If NULL, the original cell number in the reference data will be adopted.
#' @param nGroups The expected number of groups in the simulated data. Default is 1.
#' @param prop_group The proportion of cells in each group. Default is 1.
#' @param de_prop The proportion of DEGs over all genes between different simulated cell groups. Default is 0.2.
#' @param fc_group The fold change of DEGs between cell groups. Default is 2.
#' @param verbose Whether to return messages or not
#' @param seed Random seed
#' @param return_format The format of returned simulation data. Choices: list, SingleCellExperiment and Seurat.
#' @param ... Other parameters represented in scDesign, see [scDesign::design_data()]
#'
#' @importFrom purrr map
#' @importFrom utils packageVersion
#'
#' @references Li W V, Li J J. A statistical simulator scDesign for rational scRNA-seq experimental design. Bioinformatics, 2019, 35(14): i41-i50. <https://doi.org/10.1093/bioinformatics/btz321>
#'
#' @export
#'

scDesign_simulation <- function(SCST_Object,
                                cell_num = NULL,
                                nGroups = 1,
                                prop_group = 1,
                                de_prop = 0.2,
                                fc_group = 2,
                                return_format,
                                verbose = FALSE,
                                seed,
                                ...
){
  ##############################################################################
  ####                            Environment                                ###
  ##############################################################################
  if(!requireNamespace("scDesign", quietly = TRUE)){
    message("scDesign is not installed on your device...")
    message("Installing scDesign...")
    devtools::install_github("Vivianstats/scDesign")
  }
  existed_params <- list()
  existed_params[["realcount"]] <- SCST_Object@reference %>% as.matrix()
  ## nCells
  if(is.null(cell_num)){
    existed_params[["ncell"]] <- SCST_Object@cell_num
  }else{
    existed_params[["ncell"]] <- cell_num
  }

  if(length(de_prop) == 1){
    existed_params[["pUp"]] <- de_prop/2
    existed_params[["pDown"]] <- de_prop/2
  }
  if(length(de_prop) == 2){
    existed_params[["pUp"]] <- de_prop[1]
    existed_params[["pDown"]] <- de_prop[2]
  }
  if(!is.null(fc_group)){
    if(length(fc_group) == 1){
      existed_params[["fU"]] <- fc_group
      existed_params[["fL"]] <- 1/fc_group
    }
  }
  existed_params[["ngroup"]] <- nGroups
  if(nGroups > 1){
    if(length(existed_params[["ncell"]]) != existed_params[["ngroup"]]){
      ## group proportions
      if(!is.null(prop_group)){
        assert_that(nGroups == length(prop_group))
        existed_params[["ncell"]] <- proportionate(
          number = existed_params[["ncell"]],
          result_sum_strict = existed_params[["ncell"]],
          prop = prop_group,
          prop_sum_strict = 1,
          digits = 0)
      }else{
        existed_params[["ncell"]] <- proportionate(
          number = existed_params[["ncell"]],
          result_sum_strict = existed_params[["ncell"]],
          prop = rep(1/existed_params[["ngroup"]], existed_params[["ngroup"]]),
          digits = 0)
      }
    }
    extra_params <- list(...)
    existed_params <- append(existed_params, extra_params)
    argument <- formals(scDesign::design_data)
    used_argument <- change_parameters(existed_params, argument)
    if(length(used_argument[["S"]]) != used_argument[["ngroup"]]){
      used_argument[["S"]] <- rep(used_argument[["S"]], used_argument[["ngroup"]])
      assertthat::assert_that(length(used_argument[["S"]]) == used_argument[["ngroup"]],
                              msg = "The length of S must equal to nGroups")
    }
  }
  ##############################################################################
  ####                            Simulation                                 ###
  ##############################################################################
  if(verbose){
    message("Simulating datasets using scDesign")
  }
  # Seed
  set.seed(seed)
  # Estimation
  simulate_detection <- peakRAM::peakRAM(
    simulate_result <- BiocGenerics::do.call(scDesign::design_data, used_argument)
  )
  ##############################################################################
  ####                        Format Conversion                              ###
  ##############################################################################
  if(used_argument[["ngroup"]] == 1){
    counts <- simulate_result
    rownames(counts) <- paste0("Gene", 1:nrow(counts))
    colnames(counts) <- paste0("Cell", 1:ncol(counts))
    col_data <- data.frame("cell_name" = colnames(counts),
                           row.names = colnames(counts))
    row_data <- data.frame("gene_name" = rownames(counts),
                           row.names = rownames(counts))
  }else{
    counts_tmp <- simulate_result[["count"]]
    # col_data
    col_data <- data.frame("cell_name" = paste0("Cell", 1:sum(used_argument[["ncell"]])),
                           "group" = paste0("Group", unlist(purrr::map(seq_len(length(used_argument[["ncell"]])),
                                                                       function(x){
                                                                         rep(as.character(x), ncol(counts_tmp[[x]]))
                                                                       }))))
    # Get count data together
    counts <- c()
    for(i in 1:length(counts_tmp)){
      counts <- cbind(counts, counts_tmp[[i]])
    }
    # Rename
    rownames(counts) <- paste0("Gene", 1:nrow(counts))
    colnames(counts) <- paste0("Cell", 1:ncol(counts))
    rownames(col_data) <- colnames(counts)
    # Up and down regulated genes
    up_gene <- simulate_result[["genesUp"]][[2]] %>%
      stringr::str_extract(pattern = "[0-9]+") %>%
      as.numeric()
    down_gene <- simulate_result[["genesDown"]][[2]] %>%
      stringr::str_extract(pattern = "[0-9]+") %>%
      as.numeric()

    row_data <- data.frame("gene_name" = rownames(counts),
                           "de_gene" = "no",
                           "up_down" = "no",
                           row.names = rownames(counts))
    # Row data
    row_data[up_gene, 3] <- "up"
    row_data[down_gene, 3] <- "down"
    row_data <- row_data %>%
      dplyr::mutate("de_gene" = dplyr::case_when(
        up_down == "no" ~ "no",
        TRUE ~ "yes"
      ))
  }

  # Establish SingleCellExperiment
  simulate_result <- SingleCellExperiment::SingleCellExperiment(list(counts = counts),
                                                                colData = col_data,
                                                                rowData = row_data)
  simulate_result <- data_conversion(SCE_object = simulate_result, return_format = return_format)
  ##############################################################################
  ####                           Ouput                                       ###
  ##############################################################################
  slot(SCST_Object, "estimation_time_memory") <- list("scDesign" = list("estimation_time" = NULL,
                                                                        "estimation_memory" = NULL))
  slot(SCST_Object, "estimation_result") <- list("scDesign" = NULL)
  slot(SCST_Object, "estimation_parameters") <- list("scDesign" = NULL)
  slot(SCST_Object, "simulation_time_memory") <- list("scDesign" = list("estimation_time" = simulate_detection$Elapsed_Time_sec,
                                                                        "estimation_memory" = simulate_detection$Peak_RAM_Used_MiB))
  slot(SCST_Object, "simulation_result") <- list("scDesign" = simulate_result)
  slot(SCST_Object, "simulation_parameters") <- list("scDesign" = used_argument[-1])
  return(SCST_Object)
}
