#' Estimation Parameters Using SCRIP
#'
#' @param SCST_Object A SCST object generated by [SCST::Create_SCST_Object()]
#' @param verbose Whether to return messages or not
#'
#' @export
#'
#' @references Qin F, Luo X, Xiao F, et al. SCRIP: an accurate simulator for single-cell RNA sequencing data. Bioinformatics, 2022, 38(5): 1304-1311. <https://doi.org/10.1093/bioinformatics/btab824>
#'
SCRIP_estimation <- function(SCST_Object,
                             verbose = FALSE
){
  ##############################################################################
  ####                               Check                                   ###
  ##############################################################################
  if(!requireNamespace("splatter")){
    message("Installing splatter ...")
    BiocManager::install("splatter")
  }
  ref_data <- methods::slot(SCST_Object, "reference")
  seed <- methods::slot(SCST_Object, "seed")
  ##############################################################################
  ####                            Estimation                                 ###
  ##############################################################################
  if(verbose){
    message("Estimating parameters using SCRIP")
  }
  # Seed
  set.seed(seed)
  # Estimation
  estimate_detection <- peakRAM::peakRAM(
    estimate_result <- splatter::splatEstimate(as.matrix(ref_data))
  )
  ##############################################################################
  ####                           Ouput                                       ###
  ##############################################################################
  slot(SCST_Object, "estimation_time_memory") <- list("SCRIP" = list("estimation_time" = estimate_detection$Elapsed_Time_sec,
                                                                     "estimation_memory" = estimate_detection$Peak_RAM_Used_MiB))
  slot(SCST_Object, "estimation_result") <- list("SCRIP" = estimate_result)
  slot(SCST_Object, "estimation_parameters") <- list("SCRIP" = NULL)
  return(SCST_Object)
}



#' Simulate ScRNA-seq Data Using SCRIP
#'
#' @param estimated_result The SCST object after estimating parameters using [splatter::splatEstimate()]
#' @param mode SCRIP contains five different simulation modes, and you can specify which mode do you use (default is GP-trendedBCV): 1. GP-trendedBCV; 2. GP-commonBCV; 3. BGP-commonBCV 4. BP 5. BGP-trendedBCV
#' @param cell_num The expected number of simulated cells. If NULL, the original cell number in the reference data will be adopted.
#' @param gene_num The expected number of simulated genes. If NULL, the original gene number in the reference data will be adopted.
#' @param nGroups The expected number of groups in the simulated data. Default is 1.
#' @param prop_group The proportion of cells in each group. Default is 1.
#' @param de_prop The proportion of DEGs over all genes between different simulated cell groups. Default is 0.2.
#' @param de_facLoc Location (meanlog) parameter for the differential expression factor log-normal distribution. Can be a vector.
#' @param de_facScale Scale (sdlog) parameter for the differential expression factor log-normal distribution. Can be a vector.
#' @param nBatches The number of cell batches to be simulated. Default is 1.
#' @param prop_batch The proportion of cells in each batch. Default is 1.
#' @param batch_facLoc Location (meanlog) parameter for the batch effect factor log-normal distribution. Can be a vector.
#' @param batch_facScale Scale (sdlog) parameter for the batch effect factor log-normal distribution. Can be a vector.
#' @param path A Boolean symbol (TRUE or FLASE) which determines whether to simulate datasets with trajectory along the path. Default is false.
#' @param verbose Whether to return messages or not
#' @param seed Random seed
#' @param return_format The format of returned simulation data. Choices: list, SingleCellExperiment and Seurat.
#' @param ... Other parameters represented in SCRIP, see [SCRIP::SCRIPsimu()]
#'
#'
#' @references Qin F, Luo X, Xiao F, et al. SCRIP: an accurate simulator for single-cell RNA sequencing data. Bioinformatics, 2022, 38(5): 1304-1311. <https://doi.org/10.1093/bioinformatics/btab824>
#'
#' @export
#'
SCRIP_simulation <- function(estimated_result,
                             mode = "GP-trendedBCV",
                             cell_num = NULL,
                             gene_num = NULL,
                             nGroups = 1,
                             prop_group = 1,
                             de_prop = 0.2,
                             de_facLoc = 0.1,
                             de_facScale = 0.4,
                             nBatches = 1,
                             prop_batch = 1,
                             batch_facLoc = 0.1,
                             batch_facScale = 0.1,
                             path = FALSE,
                             verbose = FALSE,
                             return_format,
                             seed,
                             ...
){
  ##############################################################################
  ####                               Check                                   ###
  ##############################################################################
  if(!requireNamespace("SCRIP", quietly = TRUE)){
    message("SCRIP is not installed on your device")
    message("Installing SCRIP")
    devtools::install_github("thecailab/SCRIP")
  }
  if(nGroups != length(prop_group)){
    print_color_word("no The number of nGroups and the length of prop_group vector should be equal.")
  }
  if(nBatches != length(prop_batch)){
    print_color_word("no The number of nBatches and the length of prop_batch vector should be equal.")
  }
  parameters <- estimated_result@estimation_result$SCRIP
  extra_params <- list(...)
  candidate_parameters <- list()
  candidate_parameters[["de.prob"]] <- de_prop
  candidate_parameters[["de.facLoc"]] <- de_facLoc
  candidate_parameters[["de.facScale"]] <- de_facScale
  candidate_parameters[["batch.facLoc"]] <- batch_facLoc
  candidate_parameters[["batch.facScale"]] <- batch_facScale
  if(length(prop_group) > 1 & nGroups > 1){
    candidate_parameters[["group.prob"]] <- prop_group
  }
  if(length(prop_batch) > 1 & nBatches > 1){
    if(is.null(cell_num)){
      cell_num <- parameters@nCells
    }
    batchCells <- proportionate(number = cell_num,
                                result_sum_strict = cell_num,
                                prop = prop_batch,
                                prop_sum_strict = 1)
    candidate_parameters[["batchCells"]] <- as.numeric(batchCells)
  }
  candidate_parameters <- append(candidate_parameters, extra_params)
  parameters <- change_parameters(parameters, candidate_parameters)
  ##############################################################################
  ####                            Simulation                                 ###
  ##############################################################################
  if(verbose){
    message("Simulating datasets using SCRIP")
  }
  # Seed
  parameters <- splatter::setParam(parameters, name = "seed", value = seed)
  parameters <- splatter::setParam(parameters,
                                   name = "de.prob",
                                   value = de_prop/parameters@nGroups)
  # Simulation
  if(path){
    cat("Simulating trajectory datasets by Splat \n")
    submethod <- "paths"
    if(parameters@path.from == 0){
      parameters <- splatter::setParam(parameters,
                                       name = "path.from",
                                       value = seq(1:parameters@nGroups)-1)
    }
  }else{
    if(parameters@nGroups == 1){
      submethod <- "single"
    }else if(parameters@nGroups != 1){
      submethod <- "groups"
    }
  }
  print(mode)
  print(parameters@batch.facLoc)
  print(parameters@batch.facScale)
  simulate_detection <- peakRAM::peakRAM(
    simulate_result <- SCRIP::SCRIPsimu(data = as.matrix(estimated_result@reference),
                                        params = parameters,
                                        method = submethod,
                                        mode = mode))
  used_argument <- list()
  used_argument[["verbose"]] <- verbose
  used_argument[["parameters"]] <- parameters
  used_argument[["method"]] <- submethod
  used_argument[["seed"]] <- seed
  ##############################################################################
  ####                        Format Conversion                              ###
  ##############################################################################
  # counts
  counts <- as.matrix(SingleCellExperiment::counts(simulate_result))
  # col_data
  col_data <- as.data.frame(SummarizedExperiment::colData(simulate_result))
  if(parameters@nGroups == 1){
    col_data[, 3] <- rep("Group1", ncol(simulate_result))
  }else{
    if(path){
      col_data$Group <- stringr::str_replace_all(col_data$Group,
                                                 pattern = "Path",
                                                 replacement = "Group")
      col_data <- col_data[, -c(4, 5)]
    }else{
      col_data <- col_data[, -4]
    }
  }
  colnames(col_data) <- c("cell_name", "batch", "group")
  # row_data
  row_data <- as.data.frame(SummarizedExperiment::rowData(simulate_result))
  if(parameters@nGroups == 1){
    row_data <- data.frame("gene_name" = rownames(counts))
    rownames(row_data) <- rownames(counts)
  }else{
    group_fac <- row_data[, grep(colnames(row_data), pattern = "^DE")]
    batch_fac <- row_data[, grep(colnames(row_data), pattern = "^Batch")]
    total_sum <- rowSums(group_fac)
    de_gene <- ifelse(total_sum == parameters@nGroups, "no", "yes")
    row_data[, 2] <- de_gene
    if(path){
      row_data <- row_data[, -c(3:4, (ncol(row_data)-ncol(group_fac)+1):ncol(row_data))]
      colnames(row_data) <- c("gene_name", "de_gene", colnames(batch_fac), paste0("DEFacGroup", 1:ncol(group_fac)))
    }else{
      row_data <- row_data[, -c(3:4)]
      colnames(row_data) <- c("gene_name", "de_gene", colnames(batch_fac), colnames(group_fac))
    }
  }
  rownames(row_data) <- row_data$gene_name
  # Establish SingleCellExperiment
  simulate_result <- SingleCellExperiment::SingleCellExperiment(list(counts = counts),
                                                                colData = col_data,
                                                                rowData = row_data)
  simulate_result <- data_conversion(SCE_object = simulate_result, return_format = return_format)
  ##############################################################################
  ####                           Ouput                                       ###
  ##############################################################################
  slot(estimated_result, "simulation_time_memory") <- list("SCRIP" = list("estimation_time" = simulate_detection$Elapsed_Time_sec,
                                                                          "estimation_memory" = simulate_detection$Peak_RAM_Used_MiB))
  slot(estimated_result, "simulation_result") <- list("SCRIP" = simulate_result)
  slot(estimated_result, "simulation_parameters") <- list("SCRIP" = used_argument)
  return(estimated_result)
}
